sui_core/src/authority/authority_checkpoints.rs

```diff
@@ -0,0 +1,1229 @@
+        fp_ensure!(
+            self.next_checkpoint() == checkpoint_sequence_number,
+            SuiError::GenericAuthorityError {
+                error: format!(
+                    "Unexpected certificate, expected next seq={}",
+                    self.next_checkpoint()
+                ),
+            }
+        );
+
+        // Sign the new checkpoint
+        let signed_checkpoint = AuthenticatedCheckpoint::Signed(
+            SignedCheckpoint::new_from_summary(checkpoint, self.name, &*self.secret),
+        );
+
+        // Make a DB batch
+        let batch = self.checkpoints.batch();
+
+        // Last store the actual checkpoints.
+        let batch = batch
+            .insert_batch(
+                &self.checkpoints,
+                [(&checkpoint_sequence_number, &signed_checkpoint)],
+            )?
+            // Drop the fragments for the previous checkpoint
+            .delete_batch(
+                &self.fragments,
+                self.fragments.iter().filter_map(|(k, v)| {
+                    // Delete all keys for checkpoints smaller than what we are committing now.
+                    if v.proposer.0.checkpoint.waypoint.sequence_number
+                        <= checkpoint_sequence_number
+                    {
+                        Some(k)
+                    } else {
+                        None
+                    }
+                }),
+            )?
+            .delete_batch(&self.local_fragments, self.local_fragments.keys())?;
+
+        // Update the transactions databases.
+        let transactions: Vec<_> = contents.transactions.iter().cloned().collect();
+        self.update_new_checkpoint_inner(checkpoint_sequence_number, &transactions, batch)?;
+
+        // Try to set a fresh proposal, and ignore errors if this fails.
+        let _ = self.set_proposal();
+
+        Ok(())
+    }
+
+    /// Call this function internally to register the latest batch of
+    /// transactions processed by this authority. The latest batch is
+    /// stored to ensure upon crash recovery all batches are processed.
+    pub fn handle_internal_batch(
+        &self,
+        next_sequence_number: TxSequenceNumber,
+        transactions: &[(TxSequenceNumber, TransactionDigest)],
+    ) -> Result<(), SuiError> {
+        self.update_processed_transactions(transactions)?;
+
+        // Updates the local sequence number of transactions processed.
+        let locals = self.get_locals();
+        let mut new_locals = locals.as_ref().clone();
+        new_locals.next_transaction_sequence = next_sequence_number;
+        self.set_locals(locals, new_locals)?;
+
+        Ok(())
+    }
+
+    // TODO: this function should submit the received fragment to the
+    //       consensus algorithm for sequencing. It should also do some
+    //       basic checks to not submit redundant information to the
+    //       consensus, as well as to check it is the right node to
+    //       submit to consensus.
```

Comment:
Now the checkpoint fragment specific checks are implemented in `handle_receive_fragment`. I do not think that the checkpoint fragment specific logic can be common with the shared object logic, since the checks are entirely different, right?